crates/pallet-domains/src/lib.rs

```diff
@@ -533,44 +512,48 @@ impl<T: Config> Pallet<T> {
             trace: Vec::new(),
             trace_root: Default::default(),
         };
-        Self::apply_execution_receipt(&genesis_receipt);
+        Self::apply_new_best_receipt(&genesis_receipt);
+        // Explicitly initialize the oldest receipt number even not necessary as ValueQuery is used.
+        OldestReceiptNumber::<T>::put::<T::BlockNumber>(Zero::zero());
     }
 
     fn pre_dispatch_submit_bundle(
         signed_opaque_bundle: &SignedOpaqueBundle<T::BlockNumber, T::Hash, T::SecondaryHash>,
     ) -> Result<(), TransactionValidityError> {
         let execution_receipts = &signed_opaque_bundle.bundle.receipts;
 
-        let (_, mut best_number) = <ReceiptHead<T>>::get();
+        if !execution_receipts
+            .iter()
+            .map(|r| r.primary_number)
+            .is_sorted()
+        {
+            return Err(TransactionValidityError::Invalid(
+                InvalidTransactionCode::ExecutionReceipt.into(),
+            ));
+        }
 
-        for execution_receipt in execution_receipts {
-            let primary_number = execution_receipt.primary_number;
+        let (_, mut best_number) = <ReceiptHead<T>>::get();
 
-            // Ensure the block number of next execution receipt is `best_number + 1`.
-            if primary_number != best_number + One::one() {
-                if primary_number <= best_number {
-                    return Err(InvalidTransaction::Stale.into());
-                } else {
-                    return Err(InvalidTransaction::Future.into());
+        for receipt in execution_receipts {
+            // Non-best receipt
+            if receipt.primary_number <= best_number {
+                continue;
```

Comment:
Should we check if it is the same as the one we stored here as well as the primary block hash, similarly to the check you have below?

Reply:
We don't have to check whether it's the same as the one we already stored as the fork receipt is possible, but I don't think it's bad to check the primary block hash even if it's already checked in the validate_unsigned.